Sergey Uvarov

lightmg opened the pull request:
@Denchick

lightmg committed:
Initial homework commit

lightmg committed:
Add simple rectangles visualiser

lightmg committed:
Testing infrastructure: LayouterTestBase added

lightmg committed:
small cleanup

lightmg committed:
Add SaveResultsAttribute.cs to make possible configure when result images should be saved

lightmg committed:
First implementation of ICircularCloudLayouter and tests (WIP)

lightmg committed:
Random rectangle colors on layouter tests

lightmg committed:
Printing execution time for every test in LayouterTestBase, Disabled rendering coordinate grid in result

lightmg committed:
First working (something like it) implementation.
Some coordinates bugs fixed, another bugs added (WIP)

lightmg committed:
Fixed one-pixel intersected rectangles

lightmg committed:
RectanglesVisualiser now abstract + static factory members added

lightmg committed:
Strange thing removed in CircularCloudLayouter.cs ¯\_(ツ)_/¯

lightmg committed:
More measurements in LayouterTestBase

lightmg committed:
Tests refactoring

lightmg committed:
Changed output directory of testing files, clearing outdated test results before testing

lightmg committed:
small cleanup

lightmg committed:
Visualising words (early WIP)

Denchick commented on cs/TagsCloudVisualisationTests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,111 @@
+        }
+
+        [Test, Ignore("Dumb test to visualise things")]
+        public void TEST_TO_REMOVE() //TODO REMOVE
```
todo забыл

lightmg replied:
Эту тудуху не забыл, она тут пока что нужна, чтобы не забыть выпилить тест к "релизу"
Пока пуллреквест черновой оставил, потом убью её

Denchick commented on cs/TagsCloudVisualisation/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,43 @@
+            if (rectangles.Count == 0)
+                return Rectangle(new Rectangle(new Point(rectangleSize / -2), rectangleSize));
+
+            var position = points.First(p => !IsIntersectsWithAny(p.PlaceRectangle(rectangleSize)));
```
Хм, все-таки в этом месте лучше PlaceRectangle сделать в виде обычного метода этого класса, а не extension метода. extensions удобно использовать, например,  когда у тебя нет возможности внести изменения в существующий класс (String, Enumerable, etc) https://stackoverflow.com/questions/787324/when-do-you-use-extension-methods-ext-methods-vs-inheritance

Denchick commented on cs/TagsCloudVisualisation/Visualisation/RectanglesVisualiser.Factory.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    public delegate void VisualisationDrawer(Graphics g, Rectangle rectangle);
+
+    public partial class RectanglesVisualiser
```
Использовать partial классы в 2020 дурной тон -- сейчас они просто не нужны. Насколько я помню (а я могу ошибаться), partial классы активно использовались в c# до появления generics, т.е. капец как давно. До дженериков приходилось генерировать код автоматически, например чтобы tuple мог принимать 1,2,3 ... n параметров. [Пруф](https://softwareengineering.stackexchange.com/questions/71494/why-use-partial-classes)

Denchick replied:
Выпили все partial, пожалуйста

lightmg replied:
Случайно образовалось, раньше там другая логика была в визуалайзере и partial-class использовался для хранения публичных статических фабричных методов и private-классов, чтобы код основного класса не был засорен этим делом

Denchick commented on cs/TagsCloudVisualisationTests/LayouterTestExtensions.cs:
```diff
@@ -0,0 +1,33 @@
+            return expected;
+        }
+
+        public static ICircularCloudLayouter PutAndTest(this ICircularCloudLayouter layouter,
```
Этот метод не используется

Denchick commented on cs/TagsCloudVisualisationTests/LayouterTestExtensions.cs:
```diff
@@ -0,0 +1,33 @@
+{
+    public static class LayouterTestExtensions
+    {
+        public static Rectangle PutAndTest(this ICircularCloudLayouter layouter,
```
Этот метод лучше перенести в класс, где он используется

lightmg replied:
Какое-то время было несколько классов с тестами, артефакт остался, выпилю

Denchick commented on cs/TagsCloudVisualisationTests/LayouterTestExtensions.cs:
```diff
@@ -0,0 +1,33 @@
+            return layouter;
+        }
+
+        public static ICircularCloudLayouter Put(this ICircularCloudLayouter layouter, Size size, out Rectangle result)
```
Этот метод лучше перенести в класс, где он используется

Denchick commented on cs/TagsCloudVisualisationTests/Infrastructure/TestingHelpers.cs:
```diff
@@ -0,0 +1,28 @@
+            Directory.CreateDirectory(fullPath);
+        }
+
+        private static string EnsureDirectory(string fullPath)
```
Обычно, подобные методы уже реализованы в библиотекак, поставляемых с языком, тебе нужно лишь их [найти](https://stackoverflow.com/questions/9065598/if-a-folder-does-not-exist-create-it).

Denchick commented on cs/TagsCloudVisualisationTests/Infrastructure/LayouterTestBase.cs:
```diff
@@ -0,0 +1,132 @@
+    {
+        private TestStatus[] statusesToSaveImage;
+
+        private static string TestOutputDirectory =>
```
Насколько я понял, тебе нужно куда-то временно сохранить какие-то файлы. Для этого обычно изпользуют папку с временными файлами в твоей ОС :)
Преимущество этой папки в том, что чистить за собой файлы не надо, этим займется сама ОС.
https://stackoverflow.com/questions/944483/how-to-get-temporary-folder-for-current-user

lightmg replied:
Зачем засорять Temp винды, когда можно засорять папку bin в текущем билде?))) 
Конечно да, мне приходится чистить это из кода (всего двумя строчками), зато я в любой момент могу зайти и посмотреть результаты прогона тестов в прошлом, удобная штука, юзаю периодически

Denchick replied:
Засорян bin в продакшене не всегда получится, но в данной конкретной ситуации пойдет. В общем, на твое усмотрение тогда

Denchick commented on cs/TagsCloudVisualisation/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,43 @@
+            return Rectangle(position.PlaceRectangle(rectangleSize));
+        }
+
+        private PlacedRectangle Rectangle(Rectangle rect)
```
воу-воу-воу, полегче

1. Название метода обязательно должно быть глаголом - чуть позже у вас будет блок про чистый код, на котором вам расскажут про это
2. Метод называется Rectangle, а создает он PlacedRectangle -- вот это жесткий обман) Лучше этот метод оформить в виде конструктора класса PlacedRectangle

Denchick commented on cs/TagsCloudVisualisation/OrderedList.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace TagsCloudVisualisation
+{
+    public sealed class OrderedList<T, TKey> : ICollection<T> where TKey : IComparable<TKey>
```
В реальной работе (а мы считаем эту задачу реальной работой) примерно никогда требуется писать свои коллекции - в шарпе их хватает с лихвой. Если ты пишешь свою коллекцию, то, вероятно, ты делаешь что-то не так. Может быть тебе подойдет [SortedList](https://docs.microsoft.com/en-us/dotnet/api/system.collections.sortedlist?view=netcore-3.1)?

Denchick commented on cs/TagsCloudVisualisationTests/Infrastructure/LayouterTestBase.cs:
```diff
@@ -0,0 +1,132 @@
+
+        private static void TestWriteLine(string message) => TestContext.Out.WriteLine(message);
+
+        private class CircularCloudLayouterHolder : ICircularCloudLayouter
```
Еще в c# принято каждый класс выносить в отдельный файл.
И да, enum тоже стоит вынести в отдельный файл

lightmg replied:
Обычно так и делаю, но я стараюсь не засорять общее пространство имен классами, которые там не нужны, они банально в ReSharper/IntelliSense потом светятся и мешают. А смысл этого класса (holder-а) в том, что он виден только изнутри, в других местах не известно о его существовании. И тут либо создавать partial-класс LayouterTestBase и выносить Holder туда, либо Holder делать public/internal и тем самым делать его видимым снаружи, что будет только мешать

Denchick reviewed (commented):
Резюме

1. Нужно поправить штуки типа "а в c# так не принято"
2. Выпилить все partial классы
3. Выпилить все extension method'ы, т.к. они тебе просто не нужны
3. Каждый класс с названием BlaBlaBla должен лежать в файле BlaBlaBla.cs

👍 за нормальную историю коммитов - ты первый шпорун у меня, кто так сделал :)
👍 за необычное решение

lightmg committed:
After-review refactoring, Part 1

lightmg committed:
More testing messages!

lightmg committed:
Merge branch 'review-refactoring' into homework-drawing-words

lightmg committed:
Tests clean-up

lightmg committed:
Some layouter fixes

lightmg committed:
Bugs killed, words visualisation added, a much of refactoring of CloudVisualiser and tests

lightmg committed:
Merge branch 'homework-drawing-words' into homework

lightmg committed:
Testing infrastructure improvements

lightmg committed:
Different-sized words has arrived! OwO

lightmg committed:
Optimized default instance of Graphics in WordsLayouter.DrawWord

lightmg committed:
Words visualisation refactoring

lightmg committed:
Word frequency based text visualisation

lightmg committed:
Ugly optimizations of beautiful layouter -.-

lightmg committed:
Testing infrastructure adjustments,
Auto-generation README.md with pre-rendered samples

lightmg committed:
small code clean-up

lightmg committed:
ReadmeBuilder extracted,
Automatically generate samples from files in TestData directory,
Words limitations added to SimpleTextAnalyzer,
Two new TestData's added,
README and Samples updated

lightmg committed:
Fixed relative path problem

lightmg committed:
Fixed relative path problem & updated README.md again ._.

lightmg committed:
CircularCloudLayouter refactored

lightmg committed:
Configurable minimal rectangle size
